src/relay.cpp

```diff
@@ -358,82 +373,148 @@ void prepare_relay_config(relay_config &interface_config, int local_sock, int fi
- * @param server_sock       pointer to socket binded to link_local address for relaying server message to client
+ * @param gua_sock  socket binded to global address for relaying client message to server and listening for server message
+ * @param lla_sock  socket binded to link_local address for relaying server message to client
  *
- * @return              none
+ * @return          int
  */
-void prepare_socket(int &local_sock, int &server_sock, relay_config &config) {
+int prepare_vlan_sockets(int &gua_sock, int &lla_sock, relay_config &config) {
     struct ifaddrs *ifa, *ifa_tmp;
-    sockaddr_in6 addr = {0};
-    sockaddr_in6 ll_addr = {0};
+    sockaddr_in6 gua = {0}, lla = {0};
 
-    if ((local_sock = socket(AF_INET6, SOCK_DGRAM, 0)) == -1) {
-        syslog(LOG_ERR, "socket: Failed to create socket on interface %s\n", config.interface.c_str());
+    if ((gua_sock = socket(AF_INET6, SOCK_DGRAM, 0)) == -1) {
+        syslog(LOG_ERR, "socket: Failed to create gua socket on interface %s\n", config.interface.c_str());
+        return -1;
     }
 
-    if ((server_sock = socket(AF_INET6, SOCK_DGRAM, 0)) == -1) {
-        syslog(LOG_ERR, "socket: Failed to create socket on interface %s\n", config.interface.c_str());
+    if ((lla_sock = socket(AF_INET6, SOCK_DGRAM, 0)) == -1) {
+        syslog(LOG_ERR, "socket: Failed to create lla socket on interface %s\n", config.interface.c_str());
+        close(gua_sock);
+        return -1;
     }
 
+    evutil_make_listen_socket_reuseable(gua_sock);
+    evutil_make_socket_nonblocking(gua_sock);
+    evutil_make_listen_socket_reuseable(lla_sock);
+    evutil_make_socket_nonblocking(lla_sock);
+
     int retry = 0;
-    bool bind_addr = false;
-    bool bind_ll_addr = false;
+    bool bind_gua = false;
+    bool bind_lla = false;
     do {
         if (getifaddrs(&ifa) == -1) {
             syslog(LOG_WARNING, "getifaddrs: Unable to get network interfaces with %s\n", strerror(errno));
         }
         else {
             ifa_tmp = ifa;
             while (ifa_tmp) {
-                if (ifa_tmp->ifa_addr && (ifa_tmp->ifa_addr->sa_family == AF_INET6) && (strcmp(ifa_tmp->ifa_name, config.interface.c_str()) == 0)) {
-                    struct sockaddr_in6 *in6 = (struct sockaddr_in6*) ifa_tmp->ifa_addr;
-                    if(!IN6_IS_ADDR_LINKLOCAL(&in6->sin6_addr)) {
-                        bind_addr = true;
-                        in6->sin6_family = AF_INET6;
-                        in6->sin6_port = htons(RELAY_PORT);
-                        addr = *in6;
-                    }
-                    if(IN6_IS_ADDR_LINKLOCAL(&in6->sin6_addr)) {
-                        bind_ll_addr = true;
-                        in6->sin6_family = AF_INET6;
-                        in6->sin6_port = htons(RELAY_PORT);
-                        ll_addr = *in6;
+                if (ifa_tmp->ifa_addr && (ifa_tmp->ifa_addr->sa_family == AF_INET6)) {
```

Comment:
We probably need some check like this https://github.com/sonic-net/sonic-buildimage/blob/202012/dockers/docker-dhcp-relay/wait_for_intf.sh.j2 for GUA's.

Reply:
Yes, good point, will make the code change.